language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -386,11 +388,17 @@ genDefDataType curPkgId conName mod tpls def =
                             ] ++
                             ["};"]
                         associatedTypes =
-                          maybe [] (\key ->
-                              [ "export namespace " <> conName <> " {"
-                              , "  export type Key = " <> fst (genType (moduleName mod) (tplKeyType key)) <> ""
-                              , "}"
-                              ]) (tplKey tpl)
+                          let tT = conName
+                              tK = tT <> ".Key"
+                              tI = "typeof " <> tT <> ".templateId" in
+                          [ "export namespace " <> tT <> " {"
+                          , "  export type Key = " <> maybe "unknown" (fst . genType (moduleName mod) . tplKeyType) (tplKey tpl)
```

Comment:
I would prefer if there was no `Key` type synonym when there's no contract key for the template. Let's use `Key` for `tK` when there is a key and `undefined` when there's no key.

Reply:
PR https://github.com/digital-asset/daml/pull/5587.